File: frontend/src/services/preferences.ts
import { apiRequest } from "@/lib/api/client";

export interface ContentPreference {
  id: string;
  workspace_id: string;
  niche_slug: string | null;
  target_audience: string | null;
  content_goals: string[];
  tone: string | null;
  language: string;
  post_style: string | null;
  cta_preference: string | null;
  hashtag_policy: string | null;
  emoji_policy: string | null;
  topics_to_avoid: string[];
  preferred_post_length: string | null;
}

export interface ContentPreferenceInput {
  niche_slug?: string | null;
  target_audience?: string | null;
  content_goals?: string[];
  tone?: string | null;
  language?: string;
  post_style?: string | null;
  cta_preference?: string | null;
  hashtag_policy?: string | null;
  emoji_policy?: string | null;
  topics_to_avoid?: string[];
  preferred_post_length?: string | null;
}

export interface VoiceProfile {
  id: string;
  workspace_id: string;
  summary: string | null;
  traits: string[];
  banned_phrases: string[];
  confidence_score: number;
  sample_count: number;
}

export interface VoiceProfileInput {
  summary?: string | null;
  traits?: string[];
  banned_phrases?: string[];
  confidence_score?: number;
}

export interface WritingSample {
  id: string;
  workspace_id: string;
  user_id: string;
  title: string | null;
  body: string;
  source: string | null;
}

export interface WritingSampleInput {
  title?: string | null;
  body: string;
  source?: string | null;
}

export interface NicheProfile {
  id: string;
  slug: string;
  name: string;
  description: string | null;
  audience_types: string[];
  content_pillars: string[];
  hook_patterns: string[];
  cta_examples: string[];
  risky_claims_to_avoid: string[];
  hashtag_guidance: string[];
}

export interface NicheProfileListResponse {
  niches: NicheProfile[];
}

export interface WritingSampleListResponse {
  samples: WritingSample[];
}

export function readNicheProfiles(signal?: AbortSignal) {
  return apiRequest<NicheProfileListResponse>("/niches", {}, signal);
}

export function readContentPreferences(workspaceId: string, signal?: AbortSignal) {
  return apiRequest<ContentPreference | null>(
    `/workspaces/${workspaceId}/preferences`,
    {},
    signal,
  );
}

export function saveContentPreferences(
  workspaceId: string,
  payload: ContentPreferenceInput,
  signal?: AbortSignal,
) {
  return apiRequest<ContentPreference>(
    `/workspaces/${workspaceId}/preferences`,
    {
      method: "PUT",
      body: JSON.stringify(payload),
    },
    signal,
  );
}

export function readVoiceProfile(workspaceId: string, signal?: AbortSignal) {
  return apiRequest<VoiceProfile | null>(
    `/workspaces/${workspaceId}/voice-profile`,
    {},
    signal,
  );
}

export function saveVoiceProfile(
  workspaceId: string,
  payload: VoiceProfileInput,
  signal?: AbortSignal,
) {
  return apiRequest<VoiceProfile>(
    `/workspaces/${workspaceId}/voice-profile`,
    {
      method: "PUT",
      body: JSON.stringify(payload),
    },
    signal,
  );
}

export function readWritingSamples(
  workspaceId: string,
  signal?: AbortSignal,
) {
  return apiRequest<WritingSampleListResponse>(
    `/workspaces/${workspaceId}/writing-samples`,
    {},
    signal,
  );
}

export function addWritingSample(
  workspaceId: string,
  payload: WritingSampleInput,
  signal?: AbortSignal,
) {
  return apiRequest<WritingSample>(
    `/workspaces/${workspaceId}/writing-samples`,
    {
      method: "POST",
      body: JSON.stringify(payload),
    },
    signal,
  );
}

File: backend/app/schemas/preferences.py
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field


class ContentPreferenceBase(BaseModel):
    niche_slug: str | None = Field(default=None, max_length=120)
    target_audience: str | None = Field(default=None, max_length=500)
    content_goals: list[str] = Field(default_factory=list)
    tone: str | None = Field(default=None, max_length=120)
    language: str = Field(default="en", min_length=2, max_length=20)
    post_style: str | None = Field(default=None, max_length=120)
    cta_preference: str | None = Field(default=None, max_length=120)
    hashtag_policy: str | None = Field(default=None, max_length=120)
    emoji_policy: str | None = Field(default=None, max_length=120)
    topics_to_avoid: list[str] = Field(default_factory=list)
    preferred_post_length: str | None = Field(default=None, max_length=80)


class ContentPreferenceUpsertRequest(ContentPreferenceBase):
    pass


class ContentPreferenceResponse(ContentPreferenceBase):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    workspace_id: UUID


class VoiceProfileBase(BaseModel):
    summary: str | None = None
    traits: list[str] = Field(default_factory=list)
    banned_phrases: list[str] = Field(default_factory=list)
    confidence_score: int = Field(default=0, ge=0, le=100)


class VoiceProfileUpsertRequest(VoiceProfileBase):
    pass


class VoiceProfileResponse(VoiceProfileBase):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    workspace_id: UUID
    sample_count: int


class WritingSampleCreateRequest(BaseModel):
    title: str | None = Field(default=None, max_length=255)
    body: str = Field(min_length=20)
    source: str | None = Field(default=None, max_length=120)


class WritingSampleResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    workspace_id: UUID
    user_id: UUID
    title: str | None = None
    body: str
    source: str | None = None


class WritingSampleListResponse(BaseModel):
    samples: list[WritingSampleResponse]


class NicheProfileResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    slug: str
    name: str
    description: str | None = None
    audience_types: list[str]
    content_pillars: list[str]
    hook_patterns: list[str]
    cta_examples: list[str]
    risky_claims_to_avoid: list[str]
    hashtag_guidance: list[str]


class NicheProfileListResponse(BaseModel):
    niches: list[NicheProfileResponse]
